pixeltable/func/function.py

```diff
@@ -64,37 +73,114 @@ def display_name(self) -> str:
             return self.self_path[len(ptf_prefix):]
         return self.self_path
 
+    @property
+    def is_polymorphic(self) -> bool:
+        return len(self.signatures) > 1
+
+    @property
+    def signature(self) -> Signature:
+        assert not self.is_polymorphic
+        return self.signatures[0]
+
     @property
     def arity(self) -> int:
+        assert not self.is_polymorphic
         return len(self.signature.parameters)
 
+    def _overload_resolution(self, signature_idx: int) -> Function:
+        """
+        Return a new  `Function` that retains just the single signature at index `signature_idx`, and is otherwise
+        identical to this `Function`.
+        """
+        assert 0 <= signature_idx < len(self.signatures)
+
+        # We cache the resolutions in self.__resolutions. This ensures that each resolution is represented globally
+        # by a single Python object. We do this dynamically rather than pre-constructing them in order to avoid
+        # circular complexity in the `Function` initialization logic.
+
+        # Note that it is necessary to create a resolution even if this Function has only one signature, because
+        # there is no guarantee we have seen all the Function's @overload declarations at the time this is called.
```

Comment:
aren't we saying that we don't want that situation (interleaving creating overloads with creating FunctionCalls)? what else can be simplified if we don't have to deal with that situation?

Reply:
Good point. Fixed. Not sure that there's anything else. I did add some checks in `@overload` to ensure new overloads can't be added once either the resolutions have been cached, or a conditional_return_type has been set.
<!-- Reviewable comment -OEWnS-l54s63eUZYqJC:bruch29 -->
<!-- Sent from Reviewable.io -->